### ISSUE TITLE
Prompt was truncated. Try to shorten the prompt or increase max_embeddings_multiples 

### ISSUE THREAD

**Issue by syddharth** [ddPn08/Radiata#19]
How does one increase the max_embeddings_multiples ?

**Comment by nub2927** [comment]
> How does one increase the max_embeddings_multiples ?

so changing that value won't help as far as i can tell, 
the CLIP tokenizer has a 75 token limit 

the way automatics webui handles being able to process unlimited prompt length is by spliting the prompt into chunks to get multiple tensors ,concatenating those tensors and passing that to the unet

"Typing past standard 75 tokens that Stable Diffusion usually accepts increases prompt size limit from 75 to 150. Typing past that increases prompt size further. This is done by breaking the prompt into chunks of 75 tokens, processing each independently using CLIP's Transformers neural network, and then concatenating the result before feeding into the next component of stable diffusion, the Unet.

For example, a prompt with 120 tokens would be separated into two chunks: first with 75 tokens, second with 45. Both would be padded to 75 tokens and extended with start/end tokens to 77. After passing those two chunks though CLIP, we'll have two tensors with shape of (1, 77, 768). Concatenating those results in (1, 154, 768) tensor that is then passed to Unet without issue."

though since the tensorrt uses diffusers as the model im not sure how you would do this with a diffusers model.

**Comment by nub2927** [comment]
so under Lsmith\modules\diffusion\tensorrt\pwp.py
I tried implementing uncapped tokens however the oputput im getting is not anything tangible, 
usually random images having nothing to do with the promp or blank pics 
what i tried was changing the get prompt method of the  TensorRTPromptWeightingPipeline to 

**def get_prompts_with_weights(self, prompt: List[str], max_length: int):
        tokens = []
        weights = []
        truncated = False
        for text in prompt:
            texts_and_weights = parse_prompt(text)
            text_token = []
            text_weight = []
            for word, weight in texts_and_weights:
                token = self.tokenizer(
                    word,
                ).input_ids[1:-1]

                text_token += token
                text_weight += [weight] * len(token)
                if len(text_token) > max_length:
                    truncated = True
                    break

            if len(text_token) > max_length:
                truncated = True
                text_token = text_token[:max_length]
                text_weight = text_weight[:max_length]
            tokens.append(text_token)
            weights.append(text_weight)
        if truncated:
            print(
                "Prompt was truncated. Try to shorten the prompt or increase max_embeddings_multiples"
            )

        return tokens, weights**

to return the full amount of tokens 

and I changed the call method of to TensorRTPromptWeightingPipeline:
**@torch.no_grad()
    def __call__(
            self,
            prompt: Union[str, List[str]],
            negative_prompt: Optional[Union[str, List[str]]] = "",
            batch_size: Optional[int] = 1,
            num_images_per_prompt: Optional[int] = 1,
            **kwargs,
    ):
        if isinstance(prompt, str):
            prompt = [prompt]
        if isinstance(negative_prompt, str):
            negative_prompt = [negative_prompt]

        bos = self.tokenizer.bos_token_id
        eos = self.tokenizer.eos_token_id
        max_length = self.tokenizer.model_max_length

        prompt_tokens, prompt_weights = self.get_prompts_with_weights(
            prompt, max_length - 2
        )
        uncond_tokens, uncond_weights = self.get_prompts_with_weights(
            negative_prompt, max_length - 2
        )

        prompt_tokens=prompt_tokens[0]
        prompt_weights=prompt_weights[0]
        uncond_tokens=uncond_tokens[0]
        uncond_weights=uncond_weights[0]

        #prompt_tokens = prompt_tokens[::-1]
        #prompt_weights = prompt_weights[::-1]
        #uncond_tokens = uncond_tokens[::-1]
        #uncond_weights = uncond_weights[::-1]



        max_L = max(len(prompt_tokens), len(uncond_tokens))
        number_of_chunks = math.ceil(max_L / 75)
        text_embeddings_to_concat = []

        prompt_emds=[]
        uncond_embds=[]

        prompt_chnks=[]
        uncond_chnks=[]


        for c in range(number_of_chunks):
            lower_bound = c * 75
            upper_bound = c * 75 + 75
            # establish prompt chunk
            if lower_bound > len(prompt_tokens) or len(prompt_tokens)==0:
                prompt_tokens_chunk = []
                prompt_weights_chunk = []
            elif lower_bound == len(prompt_tokens):
                prompt_tokens_chunk = [prompt_tokens[lower_bound]]
                prompt_weights_chunk = [prompt_weights[lower_bound]]
            else:
                upper_bound_prompt = min(len(prompt_tokens), upper_bound)
                prompt_tokens_chunk = prompt_tokens[lower_bound:upper_bound_prompt]
                prompt_weights_chunk = prompt_weights[lower_bound:upper_bound_prompt]
                p=0
            # establish uncond chunk
            LL=len(uncond_tokens)
            if lower_bound > len(uncond_tokens) or len(uncond_tokens)==0:
                g=0
                uncond_tokens_chunk = []
                uncond_weights_chunk = []
            elif lower_bound == len(uncond_tokens):
                uncond_tokens_chunk = [uncond_tokens[lower_bound]]
                uncond_weights_chunk = [uncond_weights[lower_bound]]
            else:
                upper_bound_uncond = min(len(uncond_tokens), upper_bound)
                uncond_tokens_chunk = uncond_tokens[lower_bound:upper_bound_uncond]
                uncond_weights_chunk = uncond_weights[lower_bound:upper_bound_uncond]

            uncond_tokens_chunk=[uncond_tokens_chunk]
            uncond_weights_chunk=[uncond_weights_chunk]
            prompt_tokens_chunk=[prompt_tokens_chunk]
            prompt_weights_chunk=[prompt_weights_chunk]


            prompt_tokens_chunk, prompt_weights_chunk = pad_tokens_and_weights(
                prompt_tokens_chunk,
                prompt_weights_chunk,
                max_length,
                bos,
                eos,
            )
            text_input_ids = torch.tensor(
                prompt_tokens_chunk, dtype=torch.int32, device=self.device
            )

            uncond_tokens_chunk, uncond_weights_chunk = pad_tokens_and_weights(
                uncond_tokens_chunk,
                uncond_weights_chunk,
                max_length,
                bos,
                eos,
            )
            uncond_input_ids = torch.tensor(
                uncond_tokens_chunk, dtype=torch.int32, device=self.device
            )

            text_input_ids_inp = cuda.DeviceView(
                ptr=text_input_ids.data_ptr(), shape=text_input_ids.shape, dtype=np.int32
            )
            text_embeddings = self.text_encoder.infer(
                {"input_ids": text_input_ids_inp}, self.stream
            )["text_embeddings"]

            seq_len = text_embeddings.shape[1]
            text_embeddings = text_embeddings.repeat(1, num_images_per_prompt, 1)
            text_embeddings = text_embeddings.view(
                batch_size * num_images_per_prompt, seq_len, -1
            )

            max_length = text_input_ids.shape[-1]

            uncond_input_ids_inp = cuda.DeviceView(
                ptr=uncond_input_ids.data_ptr(),
                shape=uncond_input_ids.shape,
                dtype=np.int32,
            )
            uncond_embeddings = self.text_encoder.infer(
                {"input_ids": uncond_input_ids_inp}, self.stream
            )["text_embeddings"]

            prompt_emds += [text_embeddings]
            uncond_embds += [uncond_embeddings]

            prompt_chnks.extend(prompt_weights_chunk[0])
            uncond_chnks.extend(uncond_weights_chunk[0])



        text_embeddings=torch.cat(prompt_emds,dim=1)
        uncond_embeddings=torch.cat(uncond_embds,dim=1)

        prompt_tokens = prompt_tokens
        prompt_weights = prompt_weights
        uncond_tokens = uncond_tokens
        uncond_weights = uncond_weights

        seq_len = uncond_embeddings.shape[1]
        uncond_embeddings = uncond_embeddings.repeat(1, num_images_per_prompt, 1)
        uncond_embeddings = uncond_embeddings.view(
            batch_size * num_images_per_prompt, seq_len, -1
        )

        prompt_weights_chunk = torch.tensor(
            [prompt_chnks], dtype=text_embeddings.dtype, device=self.device
        )
        uncond_weights_chunk = torch.tensor(
            [uncond_chnks],
            dtype=text_embeddings.dtype,
            device=self.device,
        )

        previous_mean = (
            text_embeddings.float().mean(axis=[-2, -1]).to(text_embeddings.dtype)
        )
        text_embeddings *= prompt_weights_chunk.unsqueeze(-1)
        current_mean = (
            text_embeddings.float().mean(axis=[-2, -1]).to(text_embeddings.dtype)
        )
        text_embeddings *= (previous_mean / current_mean).unsqueeze(-1).unsqueeze(-1)
        previous_mean = (
            uncond_embeddings.float().mean(axis=[-2, -1]).to(uncond_embeddings.dtype)
        )
        uncond_embeddings *= uncond_weights_chunk.unsqueeze(-1)
        current_mean = (
            uncond_embeddings.float().mean(axis=[-2, -1]).to(uncond_embeddings.dtype)
        )
        uncond_embeddings *= (previous_mean / current_mean).unsqueeze(-1).unsqueeze(-1)

        #prompt_emds = torch.cat(uncond_embds)
        #uncond_embds = torch.cat(uncond_embds)
        #text_embeddings = torch.cat()
        text_embeddings = torch.cat([text_embeddings])

        return text_embeddings**


does returning a diffrent text embedding than expected cause things to break the way sd is done here?

**Comment by ddPn08** [comment]
Closing this issue as the project has been significantly updated. Please reopen if you still have problems.
